definition_validator/validator.cc

```diff
@@ -45,20 +45,59 @@ Signature decomposeSignature(const core::GlobalState &gs, core::MethodRef method
     return sig;
 }
 
-// This returns true if `sub` is a subtype of `super`, but it also returns true if either one is `nullptr` or if either
-// one is not fully defined. This is really just a useful helper function for this module: do not use it elsewhere.
-bool checkSubtype(const core::Context ctx, const core::TypePtr &sub, const core::TypePtr &super) {
+// This is really just a useful helper function for this module: do not use it elsewhere.
+//
+// It makes certain assuptions that it is running for the sake of computing overrides that are not
+// going to be true in other situations.
+bool checkSubtype(const core::Context ctx, core::TypeConstraint &constr, const core::TypePtr &sub,
+                  core::MethodRef subMethod, const core::TypePtr &super, core::MethodRef superMethod,
+                  core::Polarity polarity) {
     if (sub == nullptr || super == nullptr) {
+        // nullptr is just "unannotated" which is T.untyped
         return true;
     }
 
-    // type-checking these in the presence of generic type parameters is tricky, so for now we're going to punt on
-    // it. TODO: build up the machinery to type-check in the presence of type parameters!
-    if (!super.isFullyDefined() || !sub.isFullyDefined()) {
-        return true;
+    auto subOwner = subMethod.data(ctx)->owner;
+    auto superOwner = superMethod.data(ctx)->owner;
+
+    // Need this to be true for the sake of resultTypeAsSeenFrom.
+    ENFORCE(subOwner.data(ctx)->derivesFrom(ctx, superOwner));
+
+    // We're only using the TypeConstraint as a way to have an easy way to replace a `TypeVar` with
+    // a "skolem" type variable (type variable representing an unknown but specific type). In
+    // Sorbet, those skolems are SelfTypeParam types that wrap a TypeArgumentRef.
```

Comment:
This whole comment is super-helpful.

Should the `make_type` constructor for `SelfTypeParam` (or the `SelfTypeParam` constructor itself) `ENFORCE` that the `SymbolRef` it receives is, in fact, of this kind (or a type member, since it looks like we pun `SelfTypeParam` on both in various places)?